protecode/scanning.py

```diff
@@ -464,53 +490,108 @@ def result_matches(resource: cm.Resource, result: pm.AnalysisResult):
+        ):
+            query_data = protecode.util.component_artifact_metadata(
+                component=c,
+                artefact=r,
+                omit_component_version=False,
+                omit_resource_version=True,
+            )
+        else:
+            query_data = protecode.util.component_artifact_metadata(
+                component=c,
+                artefact=r,
+                omit_component_version=True,
+                omit_resource_version=True,
+            )
 
-        scans = list(protecode_client.list_apps(
+        # TODO: since we ignore all versions for some of these artifact groups we potentially request
+        # the same information multiple times. This is a quick hacked-in cache. Look into doing this
+        # properly.
+        # Note to self: adding LRU-Cache on classes is potentially a bad idea
+        meta = frozenset([(k, v) for k,v in query_data.items()])
+        scans = list(_proxy_list_apps(
+            protecode_client=protecode_client,
             group_id=group_id,
-            custom_attribs=query_data,
+            prototype_metadata=meta,
         ))
 
-        resource_id = _resource_id(resource=resource)
-        scan_results[resource_id] = scans
+        resource_group_id = _resource_group_id(resource_group=resource_group)
+        scan_results[resource_group_id] = scans
 
     return scan_results
 
 
-def _filter_resource_nodes(
+def _resource_groups(
     resource_nodes: typing.Generator[cnudie.iter.ResourceNode, None, None],
     filter_function: typing.Callable[[cnudie.iter.ResourceNode], bool],
     artefact_types=(
         cm.ResourceType.OCI_IMAGE,
         'application/tar+vm-image-rootfs',
     ),
-) -> typing.Generator[cnudie.iter.ResourceNode, None, None]:
+) -> typing.Generator[tuple[cnudie.iter.ResourceNode], None, None]:
+    '''
+    group resources of same component name and resource version name
+
+    this grouping is done in order to deduplicate identical resource versions shared between
+    different component versions.
+    '''
+
+    # artefact-groups are grouped by:
+    # component-name, resource-name, resource-version, resource-type
+    # (thus implicitly deduplicating same resource-versions on different components)
+    resource_groups: dict[tuple(str, cm.ResourceIdentity, str), list[cnudie.iter.ResourceNode]] \
+        = collections.defaultdict(list)
+
     for resource_node in resource_nodes:
         if not resource_node.resource.type in artefact_types:
             continue
 
         if filter_function and not filter_function(resource_node):
             continue
 
-        yield resource_node
+        group_id = _resource_group_id((resource_node,))
+        resource_groups[group_id].append(resource_node)
+
+    logger.info(f'{len(resource_groups)=}')
+    yield from (tuple(g) for g in resource_groups.values())
+
+
+# TODO: Hacky cache. See _retrieve_existing_scan_results
+@functools.lru_cache
+def _proxy_list_apps(
```

Comment:
why not use an explicit cache (or, as you suggested: integrate caching in bdba-client)?